File: src/main/java/Sorting/MergeSort.java
package main.java.Sorting;

public class MergeSort {

    private static void merge(int[] arr, int s, int mid, int e) {

        int[] rightHalf = new int[e - mid];
        System.arraycopy(arr, mid + 1, rightHalf, 0, e - mid);

        int lPtr = mid;
        int rPtr = rightHalf.length - 1;
        int updater = e;

        while (lPtr >= s && rPtr >= 0) {

            if (rightHalf[rPtr] >= arr[lPtr]) {
                arr[updater--] = rightHalf[rPtr--];
            } else {
                arr[updater--] = arr[lPtr--];
            }
        }

        while (rPtr >= 0) {
            arr[updater--] = rightHalf[rPtr--];
        }
    }

    private static void recursiveSort(int[] arr, int s, int e) {

        if (s >= e)
            return;

        int mid = s + (e - s) / 2;

        recursiveSort(arr, s, mid);
        recursiveSort(arr, mid + 1, e);

        merge(arr, s, mid, e);
    }

    public static void recursiveSort(int[] arr) {
        recursiveSort(arr, 0, arr.length - 1);
    }

    public static void iterativeSort(int[] arr) {

        int n = arr.length;

        for (int size = 1; size < n; size *= 2) {

            for (int left = 0; left < n; left += 2 * size) {

                int mid = Math.min(left + size - 1, n - 1);
                int right = Math.min(left + 2 * size - 1, n - 1);

                if (mid < right)
                    merge(arr, left, mid, right);
            }
        }
    }
}

File: src/main/python/Sorting/merge_sort.py
class MergeSort:

    @staticmethod
    def __recursive_sort(arr: list[int], start: int, end: int) -> None:
        if start >= end:
            return

        mid = start + (end - start) // 2

        MergeSort.__recursive_sort(arr, start, mid)
        MergeSort.__recursive_sort(arr, mid + 1, end)

        right_half = arr[mid + 1:end + 1]

        l_ptr = mid
        r_ptr = len(right_half) - 1
        updater = end

        while l_ptr >= start and r_ptr >= 0:
            if right_half[r_ptr] >= arr[l_ptr]:
                arr[updater] = right_half[r_ptr]
                r_ptr -= 1
            else:
                arr[updater] = arr[l_ptr]
                l_ptr -= 1
            updater -= 1

        while r_ptr >= 0:
            arr[updater] = right_half[r_ptr]
            r_ptr -= 1
            updater -= 1

    @staticmethod
    def recursive_sort(arr: list[int]) -> None:
        if len(arr) <= 1:
            return

        MergeSort.__recursive_sort(arr, 0, len(arr) - 1)

    @staticmethod
    def __iterative_merge(arr: list[int], start: int, mid: int, end: int) -> None:
        right_half = arr[mid + 1:end + 1]

        l_ptr = mid
        r_ptr = len(right_half) - 1
        updater = end

        while l_ptr >= start and r_ptr >= 0:
            if right_half[r_ptr] >= arr[l_ptr]:
                arr[updater] = right_half[r_ptr]
                r_ptr -= 1
            else:
                arr[updater] = arr[l_ptr]
                l_ptr -= 1
            updater -= 1

        while r_ptr >= 0:
            arr[updater] = right_half[r_ptr]
            r_ptr -= 1
            updater -= 1

    @staticmethod
    def iterative_sort(arr: list[int]) -> None:
        n = len(arr)

        size = 1
        while size < n:

            left = 0
            while left < n:

                mid = min(left + size - 1, n - 1)
                right = min(left + 2 * size - 1, n - 1)

                if mid < right:
                    MergeSort.__iterative_merge(arr, left, mid, right)

                left += 2 * size

            size *= 2